### PR TITLE
 Deprecating chip_server_cluster_attributes

### PR DESCRIPTION
**Pull request by brdandu**
Replacing a state helper such as chip_server_cluster_attributes with a stateless helper zcl_attributes_server such that we can get deprecate the stateful helpers one by one.

Github: ZAP#898

### CODE REVIEW

File: zzz_generated/darwin-framework-tool/zap-generated/cluster/Commands.h
Comment (brdandu) [inline zzz_generated/darwin-framework-tool/zap-generated/cluster/Commands.h:1]: I actually did not make any changes to this template. I think this is unrelated to my changes.

File: zzz_generated/app-common/app-common/zap-generated/attributes/Accessors.h [hunk 1]
```diff
@@ -37,22 +37,22 @@ namespace Identify {
 namespace Attributes {
 
 namespace IdentifyTime {
-EmberAfStatus Get(chip::EndpointId endpoint, uint16_t * value); // int16u
+EmberAfStatus Get(chip::EndpointId endpoint, uint16_t * value); // int16u 0
```
Comment (brdandu) [inline zzz_generated/app-common/app-common/zap-generated/attributes/Accessors.h:40]: Is array property is being outputted now

File: src/app/zap-templates/templates/app/attributes/Accessors.zapt [hunk 2]
```diff
@@ -26,7 +26,7 @@ namespace Attributes {
 {{#if_is_struct type}}
 {{else if (canHaveSimpleAccessors this)}}
 namespace {{asUpperCamelCase label}} {
-EmberAfStatus Get(chip::EndpointId endpoint, {{accessorGetterType this}} value); // {{type}} {{isArray}}
+EmberAfStatus Get(chip::EndpointId endpoint, {{accessorGetterType this}} value); // {{type}}
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/templates/app/attributes/Accessors.zapt:29]: Ah, isArray was always undefined here currently?

File: src/darwin/Framework/CHIP/templates/partials/attribute_data_callback_name.zapt [hunk 3]
```diff
@@ -14,8 +14,22 @@
       VendorId
     {{~else if_is_strongly_typed_chip_enum type~}}
         {{asUpperCamelCase parent.name preserveAcronyms=true}}Cluster{{asUpperCamelCase type preserveAcronyms=true}}
+    {{~else if (isOctetString type)~}}
+      OctetString
+    {{~else if (isCharString type)~}}
+    CharString
+    {{~else if (isArray)~}}
+    List
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/partials/attribute_data_callback_name.zapt:22]: This cannot possibly happen; isArray is the very first thing tested for above and does something _quite_ different.

File: src/darwin/Framework/CHIP/templates/partials/attribute_data_callback_name.zapt [hunk 4]
```diff
@@ -14,8 +14,22 @@
       VendorId
     {{~else if_is_strongly_typed_chip_enum type~}}
         {{asUpperCamelCase parent.name preserveAcronyms=true}}Cluster{{asUpperCamelCase type preserveAcronyms=true}}
+    {{~else if (isOctetString type)~}}
+      OctetString
+    {{~else if (isCharString type)~}}
+    CharString
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/partials/attribute_data_callback_name.zapt:20]:
```suggestion
      CharString
```

File: src/darwin/Framework/CHIP/templates/partials/attribute_data_callback_name.zapt [hunk 5]
```diff
@@ -14,8 +14,22 @@
...
+    {{~else if (isArray)~}}
+    List
+    {{~else if (is_str_equal type "float")~}}
+    Float
+    {{~else if (is_str_equal type "double")~}}
+    Double
+    {{~else if (is_str_equal type "single")~}}
+    Float
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/partials/attribute_data_callback_name.zapt:28]:
```suggestion
    {{~else if (or (is_str_equal type "float")
                   (is_str_equal type "single"))~}}
      Float
    {{~else if (is_str_equal type "double")~}}
      Double
```

File: src/darwin/Framework/CHIP/templates/partials/attribute_data_callback_name.zapt [hunk 6]
```diff
@@ -14,8 +14,22 @@
...
+    {{~else if (is_str_equal type "boolean")~}}
+    Boolean
     {{~else~}}
-      {{chipCallback.name}}
+    Int{{as_zcl_data_type_size type parent.id sizeIn='bits' roundUpToPowerOfTwo=1}}{{~#if_is_data_type_signed type parent.id~}}s{{else}}u{{/if_is_data_type_signed}}
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/partials/attribute_data_callback_name.zapt:32]:
```suggestion
      Int{{as_zcl_data_type_size type parent.id sizeIn='bits' roundUpToPowerOfTwo=1}}{{~#if_is_data_type_signed type parent.id~}}s{{else}}u{{/if_is_data_type_signed}}
```

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/zap-templates/templates/app/attributes/Accessors.zapt (hunk 2 above)
Comment (brdandu) [reply src/app/zap-templates/templates/app/attributes/Accessors.zapt:29]: Yes.

File: src/darwin/Framework/CHIP/templates/partials/attribute_data_callback_name.zapt (hunk 3 above)
Comment (brdandu) [reply src/darwin/Framework/CHIP/templates/partials/attribute_data_callback_name.zapt:22]: ok I may have picked this up from the js code. I can remove it if that is the case

**Review by bzbarsky-apple — APPROVED** [review]

**Review by andy31415 — COMMENTED** [review]
Needs MatterIDL.zapt changes as well. it also seems we want separate logic between server side and client side cluster generation (client side wants all attributes/commands/etc) where as server side seems to want to consider enabling logic.

**Comment by bzbarsky-apple** [comment]
> Needs MatterIDL.zapt changes as well

That can be a separate PR, but yes, we need to figure out how it should work.  Especially for "limited" vs "unlimited" clients ("light switch" vs "python controller").

**Review by paulr34 — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]
